compiler/fir/providers/src/org/jetbrains/kotlin/fir/scopes/impl/FirTypeIntersectionScopeContext.kt

```diff
@@ -372,81 +347,89 @@ class FirTypeIntersectionScopeContext(
         return maxVisibility
     }
 
-    private fun createIntersectionOverride(
-        mostSpecific: FirNamedFunctionSymbol,
+    private fun createIntersectionOverrideFunction(
+        mostSpecific: Collection<FirCallableSymbol<*>>,
         overrides: Collection<FirCallableSymbol<*>>,
         newModality: Modality?,
         newVisibility: Visibility,
     ): FirNamedFunctionSymbol {
-
-        val newSymbol =
-            FirIntersectionOverrideFunctionSymbol(
-                CallableId(
-                    dispatchReceiverType.classId ?: mostSpecific.dispatchReceiverClassOrNull()?.classId!!,
-                    mostSpecific.fir.name
-                ),
-                overrides
-            )
-        val mostSpecificFunction = mostSpecific.fir
+        val key = mostSpecific.first() as FirNamedFunctionSymbol
+        val keyFir = key.fir
+        val callableId = CallableId(
+            dispatchReceiverType.classId ?: keyFir.dispatchReceiverClassOrNull()?.classId!!,
+            keyFir.name
+        )
+        val newSymbol = FirIntersectionOverrideFunctionSymbol(callableId, overrides)
         FirFakeOverrideGenerator.createCopyForFirFunction(
-            newSymbol,
-            mostSpecificFunction, session, FirDeclarationOrigin.IntersectionOverride,
-            mostSpecificFunction.isExpect,
-            newDispatchReceiverType = dispatchReceiverType,
+            newSymbol, keyFir, session, FirDeclarationOrigin.IntersectionOverride, keyFir.isExpect,
             newModality = newModality,
             newVisibility = newVisibility,
+            newDispatchReceiverType = dispatchReceiverType,
+            newReturnType = if (!forClassUseSiteScope) intersectReturnTypes(mostSpecific) else null,
```

Comment:
Actually I'm thinking it should be safe to remove this condition - with the way the condition for non-trivial overrides is written, the only ways intersecting the types in class scope that I can see should also cause other errors, and fixing those would have the effect of intersecting the types manually. Consider this:

```
interface A {
  fun foo(): Any
}

interface B {
  fun foo(): String?
}

// RETURN_TYPE_MISMATCH_ON_INHERITANCE
interface C : A, B {
  // to fix:
  // override fun foo(): String
}

fun test(c: C) {
  val x: String = c.foo() // both K1 and K2 - TYPE_MISMATCH
  // but if RETURN_TYPE_MISMATCH_ON_INHERITANCE is fixed, so will be TYPE_MISMATCH;
  // so maybe intersect the types and let `test` compile just fine?
}
```

Reply:
I don't think, we should allow it, at least for now.
For me, having a need to explicitly specify the override in `C` is a good thing.

Reply:
That's the point: explicitly specifying the override would be required either way to fix RETURN_TYPE_MISMATCH_ON_INHERITANCE; the code here only affects whether the TYPE_MISMATCH is generated

Reply:
I'm not sure it 100% correct for the case of complicated Java hierarchies, so I'd leave it as is